src/lib.rs

```diff
@@ -1,20 +1,132 @@
 mod error;
-mod graph;
-mod loader;
-mod parser;
-mod reqwest_loader;
-mod resolve_import;
-
-pub use error::Error;
-pub use graph::ModuleGraph;
-pub use graph::GRAPH_VERSION;
-pub use loader::MemoryLoader;
-pub use loader::ModuleInfo;
-pub use loader::ModuleLoad;
-pub use loader::ModuleLoadFuture;
-pub use loader::ModuleLoader;
-pub use loader::ModuleSource;
-pub use loader::ModuleStream;
-pub use reqwest_loader::load_reqwest;
-pub use reqwest_loader::none_middleware;
-pub use reqwest_loader::ReqwestLoader;
+pub mod v1;
+pub mod v2;
+
+use std::pin::Pin;
+use std::sync::Arc;
+
+use futures::Future;
+use serde::Deserialize;
+use serde::Serialize;
+use tokio::io::AsyncBufReadExt;
+use tokio::io::AsyncReadExt;
+use v2::ESZIP_V2_MAGIC;
+
+pub use crate::error::ParseError;
+pub use crate::v1::EsZipV1;
+pub use crate::v2::EsZipV2;
+
+pub use deno_ast;
+pub use deno_graph;
+
+pub enum EsZip {
+  V1(EsZipV1),
+  V2(EsZipV2),
+}
+
+/// This future needs to polled to parse the eszip file.
+type EszipParserFuture<R> =
+  Pin<Box<dyn Future<Output = Result<tokio::io::BufReader<R>, ParseError>>>>;
+
+impl EsZip {
+  /// Parse a byte stream into an EsZip. This function completes when the header
+  /// is fully received. This does not mean that the entire file is fully
+  /// received or parsed yet. To finish parsing, the future returned by this
+  /// function in the second tuple slot needs to be polled.
+  pub async fn parse<R: tokio::io::AsyncRead + Unpin + 'static>(
+    reader: R,
+    maybe_length: Option<usize>,
+  ) -> Result<(EsZip, EszipParserFuture<R>), ParseError> {
+    let mut reader = tokio::io::BufReader::new(reader);
+    reader.fill_buf().await?;
+    let buffer = reader.buffer();
+    if buffer.len() >= 8 && &buffer[..8] == ESZIP_V2_MAGIC {
+      let (eszip, fut) = EsZipV2::parse(reader).await?;
+      Ok((EsZip::V2(eszip), Box::pin(fut)))
+    } else {
+      let buffer = if let Some(length) = maybe_length {
+        let mut buffer = vec![0; length];
+        reader.read_exact(&mut buffer).await?;
+        buffer
+      } else {
+        let mut buffer = Vec::new();
+        reader.read_to_end(&mut buffer).await?;
+        buffer
+      };
+      let eszip = EsZipV1::parse(&buffer)?;
+      let fut = async move { Ok::<_, ParseError>(reader) };
+      Ok((EsZip::V1(eszip), Box::pin(fut)))
+    }
+  }
+
+  pub fn get_module(&self, specifier: &str) -> Option<Module> {
```

Comment:
It's not entirely clear to me what will happen if the future has not completed yet; would I get `None`, but once the future finishes I might get `Some(Module)`? A docstring would be useful here

Reply:
No, the resolving of the parser future has no impact on the result of this call. The parser future only affects `Module#source` and `Module#source_map`